src/services/authorization.rs

```diff
@@ -1,199 +1,142 @@
 //! Authorization service.
 use std::sync::Arc;
 
+use casbin::prelude::*;
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
+
 use super::user::Repository;
 use crate::errors::ServiceError;
 use crate::models::user::{UserCompact, UserId};
 
+#[derive(Debug, Clone, Serialize, Deserialize, Hash)]
 pub enum ACTION {
     AddCategory,
     DeleteCategory,
     GetSettings,
     GetSettingsSecret,
     AddTag,
     DeleteTag,
+    DeleteTorrent,
+    BanUser,
 }
 
 pub struct Service {
     user_repository: Arc<Box<dyn Repository>>,
+    casbin_enforcer: Arc<CasbinEnforcer>,
 }
 
 impl Service {
     #[must_use]
-    pub fn new(user_repository: Arc<Box<dyn Repository>>) -> Self {
-        Self { user_repository }
+    pub fn new(user_repository: Arc<Box<dyn Repository>>, casbin_enforcer: Arc<CasbinEnforcer>) -> Self {
+        Self {
+            user_repository,
+            casbin_enforcer,
+        }
     }
 
+    ///Allows or denies an user to perform an action based on the user's privileges
+    ///
     /// # Errors
     ///
     /// Will return an error if:
-    ///
-    /// - There is not any user with the provided `UserId` (when the user id is some).
+    /// - There is no user id found in the request
+    /// - The user id is not found in the database
     /// - The user is not authorized to perform the action.
-    pub async fn authorize(&self, action: ACTION, maybe_user_id: Option<UserId>) -> Result<(), ServiceError> {
-        match action {
-            ACTION::AddCategory
-            | ACTION::DeleteCategory
-            | ACTION::GetSettings
-            | ACTION::GetSettingsSecret
-            | ACTION::AddTag
-            | ACTION::DeleteTag => match maybe_user_id {
-                Some(user_id) => {
-                    let user = self.get_user(user_id).await?;
-
-                    if !user.administrator {
-                        return Err(ServiceError::Unauthorized);
-                    }
 
+    pub async fn authorize(&self, action: ACTION, maybe_user_id: Option<UserId>) -> std::result::Result<(), ServiceError> {
+        match maybe_user_id {
+            Some(user_id) => {
+                // Checks if the user found in the request exists in the database
+                let user_guard = self.get_user(user_id).await?;
+
+                //Converts the bool administrator value to a string so the enforcer can handle the request and match against the policy file
+                let role = if user_guard.administrator { "admin" } else { "guest" };
```

Comment:
Hi @mario-nt I would use an enum for roles:

```rust
enum Role {
   Unknown, // We don't know the role because the user is not authenticated (guest).
   Member, // It's an uploader in the current version
   Admin,
}
```

Maybe in the future:

```rust
enum Role {
   Unknown,
   Member,
   Uploader,
   Moderator,
   Admin,
}
```

I've been thinking `guest` is not a "pure" role. It's a kind of access level.

```rust
enum AccessLevel {
   Guest,
   Authenticated, // or LoggedIn
}
```

We could have others like anonymous access to a given resource via a token.

cc @da2ce7